add "systemctl reboot --boot-loader-menu=" and "systemctl reboot --boot-loader-entry="

poettering opened the pull request:
As requested by @jwrdegoede. 

Fixes: #10250 #9896 #10546

(This includes a commit made from a patch from @henryptung in #10250, with attribution patched in)

poettering commented:
Also fixes: #10546

jwrdegoede commented:
Hi,

Phew that is one big patch set. I've given all the patches a thorough review and overall things look good to me.

Thank you for doing this, I think it will be really nice to have this functionality available.

I've one concern and one question about the proposed API for interfacing with bootloaders not using the suggested EFI variables. First the concern:

For such bootloaders, the list of possible boot entries returned by the BootLoaderEntries dbus call is stored in the SYSTEMD_BOOT_LOADER_ENTRIES variable. But this gets set once before logind starts and then any changes after wards, e.g. dnf installing a new kernel and removing the oldest kernel, will not be reflected in there. To me it seems better if SYSTEMD_BOOT_LOADER_ENTRIES is just a boolean and the actual list is stored in a /run/systemd/systemd_boot_loader_entries file. With a comment in the documentation that bootloader(-scripts) must update this atomically using the standard unlink + rename pattern.

Now the question, at first I was thinking why you did not just set both the EFI variable and write one of the /run/systemd/systemd_boot* files instead of messing with the environment variables. But after taking a good look at the patches I realize that that way implementing the Get and Can calls would be quite hard and indeterministic. So I understand why you are using the env variables to enable the paths for bootloaders not using the EFI env. vars.

So now I'm wondering how this will work when an user installs e.g. Fedora with grub, with our grub scripts/glue using this functionality and then the user wants to switch to systemd-boot. Because now we have some drop-in files in place to set the environment variables, which means the EFivars will not be used.  So I guess that the answer here is that those drop-in files will be part of say grub-efi-x64 and that the user has to remove the grub-efi-x64 pkg when switching to systemd-boot, right ?

Talking about those drop-in files it is not entirely clear to me how I can drop an extra systemd cfg file to set environment variables for an existing service, can you give an example of this?

###

Last I've one small remark on the code, specifically you use this construct in various places:
```
r = getenv_bool("...")
if (r == -ENXIO) {
        ...
        ...
        ...
} else if (r < 0)
        log_warning_errno(r, "Failed to parse ...: %m");
else if (r > 0) {
        ...
        ...
        ...
}
```

I don't know what the coding style for systemd is, but it looks very kernel coding style like, in the kernel the rule is that if you use braces in one part of an if-else then you must also use them elsewhere, so personally I would expect and prefer this to look like this:

```
r = getenv_bool("...")
if (r == -ENXIO) {
        ...
        ...
        ...
} else if (r < 0) {
        log_warning_errno(r, "Failed to parse ...: %m");
} else if (r > 0) {
        ...
        ...
        ...
}
```

But that might be just me, so feel free to ignore this.

poettering commented:
> stored in the SYSTEMD_BOOT_LOADER_ENTRIES variable. But this gets set once before logind starts and then any changes after wards, e.g. dnf installing a new kernel and removing the oldest kernel, will not be reflected in there. To me it seems better if SYSTEMD_BOOT_LOADER_ENTRIES is just a boolean and the actual list is stored in a /run/systemd/systemd_boot_loader_entries file. With a comment in the documentation that bootloader(-scripts) must update this atomically using the standard unlink + rename pattern.

Well, I thought about that but opted against it. My thinking about that was that if we make stuff appear in the file system in a new file format to be defined, then we should much rather just expect people to use the boot loader spec for that on disk format. I mean, we already have parsers for that in place, it would just be matter of hooking those up to logind, too. I mean, why define a new on-disk enumeration for boot menu items if there already is one?

Anyway, for the time being I didn't hook that up, but I think a follow-up PR could very well add that too.

jwrdegoede commented:
On 31-10-18 12:19, Lennart Poettering wrote:
>     stored in the SYSTEMD_BOOT_LOADER_ENTRIES variable. But this gets set once before logind starts and then any changes after wards, e.g. dnf installing a new kernel and removing the oldest kernel, will not be reflected in there. To me it seems better if SYSTEMD_BOOT_LOADER_ENTRIES is just a boolean and the actual list is stored in a /run/systemd/systemd_boot_loader_entries file. With a comment in the documentation that bootloader(-scripts) must update this atomically using the standard unlink + rename pattern.
> 
> Well, I thought about that but opted against it. My thinking about that was that if we make stuff appear in the file system in a new file format to be defined, then we should much rather just expect people to use the boot loader spec for that on disk format. I mean, we already have parsers for that in place, it would just be matter of hooking those up to logind, too. I mean, why define a new on-disk enumeration for boot menu items if there already is one?
> 
> Anyway, for the time being I didn't hook that up, but I think a follow-up PR could very well add that too.

So what you are saying is that in the future an option would be to say set SYSTEMD_BOOT_LOADER_ENTRIES to a special "bls"
value (or use a new environment variable) and then logind will directly get the entries from the bls files, right?

That sounds like a good idea to me. Although I guess we would then also need a way to select say chainload
Windows through the bls file (although that is becoming less useful with TPM2 and bitlocker requiring that
the Windows bootloader is not chained but rather loaded directly).

poettering commented:
> So what you are saying is that in the future an option would be to say set SYSTEMD_BOOT_LOADER_ENTRIES to a special "bls" value (or use a new environment variable) and then logind will directly get the entries from the bls files, right?

No, I'd assume we'd pick that up automatically, just like the EFI var, and combine them. And if you do specify the env var you'd turn both off and override it with your own.

> That sounds like a good idea to me. Although I guess we would then also need a way to select say chainload Windows through the bls file (although that is becoming less useful with TPM2 and bitlocker requiring that the Windows bootloader is not chained but rather loaded directly).

the boot loader spec works fine for chainloading windows. Windows provides its boot loader as a properly signed EFI binary, and you can easily define a boot loader spec drop-in with the `efi` stanza to load something like that. Same for MacOS.

That said, a smart boot loader (such as sd-boot) doesn't bother with boot loader spec files for that, and just autodetects windows/macos during boot, to make things more robust. Checking for this "offline" is just another grub misfeature if you ask me...

Really, this stuff should just work, and fully automatically, and hence the boot loader should be the right place to detect things and just work.

poettering commented:
> So now I'm wondering how this will work when an user installs e.g. Fedora with grub, with our grub scripts/glue using this functionality and then the user wants to switch to systemd-boot.

I am not too concerned about this. This is not something we have to make a particularly smooth transition. it's ok to suggest the user to remove the grub package and reload logind in that case I htink. replacing your boot loader is something only very technical people should do, and hence such a requirement is not too much I think

poettering commented:
regarding the getenv_bool() handling coding style. We have the general rule that {} should not be used for single-line if blocks. See CODING_STYLE in our repo for details about that.

So yes, we mostly follow kernel coding style, except for that, for not using tabs and for opening { on the same line as the function name

jwrdegoede commented:
Ok (for all of your remarks / answers)

klausenbusk commented on docs/ENVIRONMENT.md:
```diff
@@ -86,6 +86,43 @@ systemd-logind:
+  functionality is turned off entirely. If set to true instead of requesting a
+  reboot into a specific boot loader entry through EFI a file
+  `/run/systemd/reboot-to-boot-loader-entry` is created whenever this is
+  requested. The fail contains the requested boot loader entry identifier. This
```
Typo: `fail` -> `file`

klausenbusk commented on docs/ENVIRONMENT.md:
```diff
@@ -86,6 +86,43 @@ systemd-logind:
+  the boot loader menu. Takes a boolean. If set to false the functionality is
+  turned off entirely. If set to true instead of requesting a reboot into the
+  boot loader menu through EFI a file `/run/systemd/reboot-to-boot-loader-menu`
+  is created whenever this is requested. The fail contains the requested boot
```
Typo: `fail` -> `file`

poettering commented:
I have now split out #10804 #10805 #10806 from this PR, containing hopefully the not controversial parts. It would be great to get those merged first. PTAL.

poettering commented:
> Unless I'm completely misunderstanding something, the two branches (efi and non-efi) have the same problem: they assume a static list of boot entries, as grokked by the boot loader during last boot (in the efi case), or as grokked by something before the last time logind was started (non-efi case). 

Well, my assumption is that they aren't actually that dynamic after all. Only during package upgrades they might change, i.e. at a point in time package managers restart a lot of daemons anyway.

> The second variant is even weirder, because it'd require some kind of generator to run in early boot. We definitely don't want to query available boot loader entries before logind has been started. 

We don't? logind runs in late boot. Everything is mounted at that point already, hence sounds completely OK to generate some snippet (maybe EnvironmentFile= or so) beforehand.

> And we certainly don't want to restart logind to tell it about boot loader entries. Right now restarting logind causes a crash of the graphical sessions, but even it that were fixed, writing something to the unit file and restarting the service doesn't seem to be the way to do this.

Well, I'd claim that's a bug in X that should be fixed really. There has been a lot of work been done ensuring that logind doesn't lose state and open fds when restarting, this is of course all not worth much if X doesn't catch up on this... :-(

> As mentioned in the comments below, the list of entries one might want to reboot into is live, and variables (of either kind) are not the right medium to contain this information. It needs to be dynamic. Not all details have to be fully fleshed out in this PR, but there needs to be some general understanding how things will work in the future. Essentially, only an approach in which logind "reaches out" to query the list of entries is possible. If this is supposed to be pluggable, then some generator executables which run on demand seem the easiest option.

I am not convinced this is really necessary. I mean, package upgrades on Linux generally result in daemon reloads, this is just one more. Moreover I consider the external env var thing a stopgap at most.

I mean, I think we should eventually hook up bootspec.c with this, so that the list of boot menu items is gathered from that, but besides that I don't think we should add more data sources for this. I guess we could add a scheme where other boot loaders which really don't want to do the boot loader spec natively can just drop some "fake" directory with these entries somewhere and we support that. 

Or to say this differently: I think we should really push people towards standardizing on the boot loader spec. If we add another way to supply these entries, it really should be based on the boot loader spec as plug-in mechanism I think.

Or to say this differently, I think in the long run we want:

1. Entries from the env var, override everything
2. entries from the boot loader spec are primary source of data
3. augmented with data from the EFI variables (to add in things like auto-detected windows or MacOS)

Of these, this patch implements 1+2. We can add 3 in a follow-up PR. If 3rd party boot loaders pick 1, 2, or 3 as the way to hook themselves to this logic is entirely up to them. 

> There's also a usability issue: bootctl does some nice matching, and the new code in this PR requires an an exact match, which is not nice for for the user. This is OK for the first iteration, as long as there's a plan how to provide such "fuzzy" matching later. I think it'd be better to make this matching on the client side, i.e. logind provides a list of entries, and the client picks one based on what the user requests. I expect `systemctl reboot --entry=<arg>` to take the same arguments that `bootctl set-default` does, possibly with some additional auto-generated ones.
 
> Now that I think about this more, [bd2865c](https://github.com/systemd/systemd/commit/bd2865caf6e730fa26807b749b1c3269943df8e2) is also problematic — it doesn't indicate that those entries might be stale. So the logic to list entries reported by the boot loaded needs to be adjusted to mark such entries somehow.

Well, but that's hard... after all the boot loader has a slightly different view of the world than we do.

Anyway, regardless if we add bootspec.c suppotr to all of this, I think making all of this overridable with env vars is certainly desirable. Maybe not as only source of configuration, but other sources can certainly be added later on, no doubt. I'd see this independently of this PR however...

poettering commented on man/systemctl.xml:
```diff
@@ -613,11 +613,30 @@
+          <para>When used with the <command>reboot</command> command, indicate to the system's boot loader to boot into
+          a specific boot loader entry on the following boot. Takes a boot loader entry identifier as argument, or
+          <literal>help</literal> in order to list available entries. Note that not all boot loaders support this
+          functionality.</para>
```
Hmm, dbus property are not supposed to propagate errors, gbdus and things like don't expect that. I figure we should catch this client side and simply generate an error if the list is empty.

poettering commented on src/systemctl/systemctl.c:
```diff
@@ -3417,27 +3423,75 @@ static int logind_prepare_firmware_setup(void) {
+                        "org.freedesktop.login1",
+                        "/org/freedesktop/login1",
+                        "org.freedesktop.login1.Manager",
+                        "SetRebootToBootLoaderEntry",
```
the reason I named it the way it is named is that we already have `RebootToFirmwareSetup()`, and I wanted to follow the same naming regime, so that we now have:

1. `SetRebootToFirmwareSetup()`
2. `SetRebootToBootLoaderMenu()`
3. `SetRebootToBootLoaderEntry()`

keszybz replied:
Yeah, but those names are not very accurate. I doesn't matter *too* much, but those are much more likely to be used frequently, so we should pick good names.

poettering commented:
I force pushed a rebased version now. It's much shorter now of course, since all the split-out stuff got merged.

I addresses the points @keszybz made, except as indicated in the discussions above (i.e. all the stuff i have not marked resolved). PTAL!

Again, I think we should add bootspec.c support later on, plus a way how 3rd party stuff can redirect where we read these snippets from with another env var, so that they can influence dynamically what we consider valid entries to be.

poettering commented:
I force pushed a new version now. it's rebased on current git (i.e. on top of the recent bootspec mega-PR).

The big change is that instead of reading the boot entries from an env var it simply uses the bootspec stuff for everything: if a compliant boot loader is there things will just work, and if not, then other stuff can generate synthetic entries in /run/boot-loader-entries/ instead. This means we only support a single format for informing us about boot entries, and don't have to support multiple. And this scheme also works for the "systemctl kexec" stuff.

So this means that whenever some foreign boot loader installer script wants to inform us for logind/systemctl about boot loader entries it can do that by simply populating that dir with snippets and symlinks. Or it can just support the boot loader spec properly in the ESP or the XBOOTLDR partition which of course makes everything a ton easier. So it's a gentle push towards boot loaders to support the spec natively, so that they don't have to synthesize these files all the time.

yuwata commented:
CI fails with the following:
```
../src/boot/bootctl.c:1178:31: error: unused variable 'found_by_loader' [-Werror,-Wunused-variable]
        _cleanup_free_ char **found_by_loader = NULL;
                              ^
```

poettering commented:
force pushed a new version. hopefully CI passes now. ptal!

poettering commented on src/systemctl/systemctl.c:
```diff
@@ -3473,7 +3475,11 @@ static int logind_check_inhibitors(enum action a) {
 #endif
 }
 
-static int logind_prepare_firmware_setup(void) {
```
embedded boards probably don't do UEFI anyway, no?

(but yeah, if this is really an issue, let's see if people notice, and if they do we can maybe add this back as fallback)

poettering commented on src/boot/bootctl.c:
```diff
@@ -311,13 +311,15 @@ static int status_variables(void) {
 
-        printf("        title: %s%s%s%s%s%s\n",
+        printf("        title: %s%s%s%s%s%s\n"
+               "         type: %s\n",
```
hmm, yeah, i tried to come up with a better name, but I always came back to just calling them "conf", since that's also the suffix of the boot loader entry type 1 files. (maybe we should have called them .entry, or .boot or so, but i guess that's too late now)

poettering commented on src/login/logind-dbus.c:
```diff
@@ -2542,6 +2545,457 @@ static int method_can_reboot_to_firmware_setup(
+
+                        r = write_string_file_atomic_label("/run/systemd/reboot-to-loader-menu", buf);
+                        if (r < 0)
+                                return r;
```
so, i though about this but i decided not to do that in the end, since they aren't actually competing. i.e. let's say your use reboot-to-firmware-setup, reboot-to-boot-menu and reboot-to-boot-entry at the same time. if you do, then this currently means, and i think in all cases really to first boot into the setup, then show the boot menu with a timeout and select the specified entry in case the timeout is hit. Or in other words, I think it actually might make sense to combine them  in real life. Real life usecase: consider an installer type thingy: you run it and it would like to enroll a secureboot key, let's say. It can't though because the system isn't in secureboot mode. In this case it could tell the user "hey, i will reboot you to firmware setup now, please put the systemin enrollment mode there, then reboot". And then it could set all three (or at least two) of the options, to make sure that the firmware setup is entered and afterwards the right menu item booted into.

keszybz replied:
Ack. I'm not 100% convinced. At least in my testing, I found it rather confusing. But let's leave that for later, we can certainly adjust it based on initial feedback.

poettering commented:
Force pushed a new version, with the minimal changes @keszybz suggested applied, see my comments above. Taking liberty to set green label, since @keszybz already set it, and the changes are really minimal.
